include/wil/property.h

```diff
@@ -0,0 +1,226 @@
+
+      template<typename... TArgs>
+      auto invoke(TArgs&&... args) {
+        return m_handler(args...);
+      }
+    private:
+      winrt::event<T> m_handler;
+    };
+}
+
+/**
+ * @brief A default event handler that maps to [Windows.Foundation.EventHandler](https://docs.microsoft.com/uwp/api/windows.foundation.eventhandler-1).
+ * @tparam T The event data type.
+*/
+template<typename T>
+struct simple_event : details::event_base<winrt::Windows::Foundation::EventHandler<T>> {};
+
+/**
+ * @brief A default event handler that maps to [Windows.Foundation.TypedEventHandler](https://docs.microsoft.com/uwp/api/windows.foundation.typedeventhandler-2).
+ * @tparam T The event data type.
+ * @details Usage example:
+ * @code
+ *         // In IDL, this corresponds to:
+ *         //   event Windows.Foundation.TypedEventHandler<ModalPage, String> OkClicked;
+ *         wil::typed_event<MarkupSample::ModalPage, winrt::hstring> OkClicked;
+ * @endcode
+*/
+template<typename TSender, typename TArgs>
+struct typed_event : details::event_base<winrt::Windows::Foundation::TypedEventHandler<TSender, TArgs>> {};
+
+#endif
+#pragma endregion
+
+#pragma region INotifyPropertyChanged helpers
+#ifdef WINRT_Windows_UI_Xaml_Data_H
+
+/**
+ * @brief Helper base class to inherit from to have a simple implementation of [INotifyPropertyChanged](https://docs.microsoft.com/uwp/api/windows.ui.xaml.data.inotifypropertychanged).
+ * @tparam T CRTP type
+ * @details When you declare your class, make this class a base class and pass your class as a template parameter:
+ * @code
+ * struct MyPage : MyPageT<MyPage>, wil::notify_property_changed_base<MyPage> {
+ *      wil::property_with_notify<int> MyInt;
+ *
+ *      MyPage() : INIT_PROPERTY(MyInt, 42) { }
+ * };
+ * @endcode
+*/
+template<typename T, 
+    typename Xaml_Data_PropertyChangedEventHandler = winrt::Windows::UI::Xaml::Data::PropertyChangedEventHandler
+    // , typename = std::enable_if_t<std::is_convertible_v<T, winrt::Windows::Foundation::IInspectable>>
+>
+struct notify_property_changed_base {
+public:
+  using Type = T;
+  auto PropertyChanged(Xaml_Data_PropertyChangedEventHandler const& value) {
+    return m_propertyChanged.add(value);
+  }
+  void PropertyChanged(winrt::event_token const& token) {
+    m_propertyChanged.remove(token);
+  }
+  Type& self() {
+    return *static_cast<Type*>(this);
+  }
+
+  /**
+   * @brief Raises a property change notification event
+   * @param name The name of the property
+   * @return
+   * @details Usage example\n
+   * C++
+   * @code
+   * void MyPage::DoSomething() {
+   *  // modify MyInt
+   *  // MyInt = ...
+   *
+   *  // now send a notification to update the bound UI elements
+   *  RaisePropertyChanged(L"MyInt");
+   * }
+   * @endcode
+  */
+  auto RaisePropertyChanged(std::wstring_view name) {
+    return m_propertyChanged(self(), Xaml_Data_PropertyChangedEventHandler{ name });

```

Comment:
Could consider enabling this code to use deducing this when available.

Reply:
do you mean returning *this or some sfinae magic to remove the need for self(), or something else? if re: removing self, you still need it when deducing this isn't available so having the one impl is simpler and clearer.